Adding 'contains' function to wcs package

ceb8 opened the pull request:
Adding function to WCS class that checks if a given SkyCoord is in the footprint described by the wcs object. Function returns True is the coordinate is in the footprint, false otherwise.

Use as:
```
w = wcs.WCS(<some header>)
w.contains(SkyCoord(254,2,unit='deg'))
```

pllim commented on CHANGES.rst:
```diff
@@ -97,6 +97,8 @@ astropy.wcs
   and the `WCSTools documentation <http://tdc-www.harvard.edu/software/wcstools/wcstools.multiwcs.html>`_.
   [#6990]
 
+- Add ACS.contains function to check if the wcs footprint contains a given sky coordinate. [#7273]
```
Typo. And "wcs" -> "WCS" (it is acronym).

pllim commented on astropy/wcs/tests/test_wcs.py:
```diff
@@ -1054,3 +1055,50 @@ def test_keyedsip():
+    hasCoord = w.contains(SkyCoord(24,2,unit='deg'))
+    assert hasCoord == False
+
+    
```
I foresee a PEP8 failure here for extra blank line. :wink:

pllim commented on astropy/wcs/wcs.py:
```diff
@@ -57,6 +57,7 @@
 from ..utils.compat import possible_filename
 from ..utils.exceptions import AstropyWarning, AstropyUserWarning, AstropyDeprecationWarning
 
+
```
Extra blank line here unnecessary.

pllim commented:
@bsipocz , do you think this needs a "What's New" entry?

eteq commented:
FYI, @ceb8 - the test failures look to be because many of the lines in the test have an extra whitespace at the end.  I think if you remove those the tests should pass.

ceb8 commented:
@eteq Thanks! Yes, I forgot that autopep8 doesn't fix trailing whitespace. 😛

nden commented on astropy/wcs/wcs.py:
```diff
@@ -977,13 +987,20 @@ def write_dist(num, cpdis):
+            header[str('CDELT2')] = (cpdis.cdelt[1],
+                                     'Coordinate increment along axis')
+            image.ver = int(hdulist[0].header[str(
+                '{0}{1:d}.EXTVER').format(d_kw, num)])
```
Why are these changes and other similar changes necessary? What makes this section better than the way it was before? It looks like autopep8 was run on this file and it does funny things. To me this is just noise and hard to review. Next time I edit the file my settings may try to revert them. Sorry, but I prefer not to have these type of changes. Is it possible to only do the necessary PEP8 changes?

bsipocz commented:
@pllim - I don't think this is big enough feature to have a separate what's new, but if someone thinks otherwise, please add the label.

eteq commented on CHANGES.rst:
```diff
@@ -97,6 +97,8 @@ astropy.wcs
   and the `WCSTools documentation <http://tdc-www.harvard.edu/software/wcstools/wcstools.multiwcs.html>`_.
   [#6990]
 
+- Add WCS.contains function to check if the WCS footprint contains a given sky coordinate. [#7273]
```
Minor nitpick: If you render the function as 
```
``WCS.contains()``
```
it ends up looking a bit more like a function call when rendered in the html changelog.

eteq commented on astropy/wcs/tests/test_wcs.py:
```diff
@@ -1054,3 +1055,48 @@ def test_keyedsip():
+    """
+
+    f = io.StringIO(header)
+    header = fits.Header.fromtextfile(f)
```
A trick that will make the "fake a file" technique unnecessary -
 you can replace these two lines with just ``header  = fits.Header.fromstring(header.strip(),'\n')``

ceb8 replied:
@eteq Brilliant! I initially wanted to use the fromstring function but was getting massive errors.

eteq commented on astropy/wcs/wcs.py:
```diff
@@ -3068,6 +3068,34 @@ def _as_mpl_axes(self):
+        Returns
+        -------
+        response : bool
+           True means the wsc footprint contains the coordinate, False means it does not.
```
typo: wsc -> WCS?

eteq commented on astropy/wcs/wcs.py:
```diff
@@ -3068,6 +3068,34 @@ def _as_mpl_axes(self):
+        Parameters
+        ----------
+        coord : `~astropy.coordinates.SkyCoord`
+            The coordinate to check if it is withing the wcs coordinate.
```
typo: `withing` -> `within`

eteq commented on astropy/wcs/wcs.py:
```diff
@@ -3068,6 +3068,34 @@ def _as_mpl_axes(self):
+            #  Suppress warnings since they just mean we didn't find the coordinate
+            warnings.simplefilter("ignore")
+            try:
+                x,y = coord.to_pixel(self, mode=mode)
```
Perhaps you should add the `origin` keyword here?  Or alternatively, replace ``mode=mode`` with just ``**kwargs`` and note in the docstring that the kwargs get passed into `to_pixel`?

eteq commented on astropy/wcs/wcs.py:
```diff
@@ -3068,6 +3068,34 @@ def _as_mpl_axes(self):
+        ----------
+        coord : `~astropy.coordinates.SkyCoord`
+            The coordinate to check if it is withing the wcs coordinate.
+
```
missing a parameter for `mode` (although see my comment below that you might find it simpler to just use ``**kwargs``)

nden commented:
@ceb8 Thanks for reverting the PEP 8 changes.

I also like the concept. I think that the function name is too general for the functionality. Since this deals only with sky coordinates, a more specific function name will be clearer. Something like `contains_sky_coordinate` or `contains_sky_location` but I am open to other suggestions. 

In addition, I think it should be able to work with numbers only (i.e. no SkyCoord objects) and with arrays - both numerical and SkyCoord arrays. If the input is an array the function will return a bool array.
@eteq What do you think about this?

astrofrog commented:
Just a quick question - doesn't having a contains method assume that the data size is part of the public API? (I thought naxis1/naxis2 were private, or did we make them public again?)

nden commented:
@astrofrog Good point. They are still private.

ceb8 commented:
@astrofrog @nden  Since this is a class method shouldn't it be OK to access private variables?

saimn commented:
> Just a quick question - doesn't having a contains method assume that the data size is part of the public API? (I thought naxis1/naxis2 were private, or did we make them public again?)

I also remember a long discussion about the need to have public attributes for the shape:  #5461
This probably needs a decision before adding a method which requires these attributes.

eteq commented:
Re: @nden's https://github.com/astropy/astropy/pull/7273#issuecomment-372426316
> In addition, I think it should be able to work with numbers only (i.e. no SkyCoord objects) and with arrays - both numerical and SkyCoord arrays. If the input is an array the function will return a bool array.

I disagree with this.  The point of `SkyCoord` is that it gives you a way to homogenously represent coordinates and you can leave it to `SkyCoord` to do the business of translating arrays/strings/whatever into something that's actually a well defined sky coordinate. I think internally you'd end up having to do something like ``SkyCoord(ra=arr1*u.deg, dec=arr2*u.deg)`` *anyway*, so why not just say you have to do it directly?

Re: 
> Something like contains_sky_coordinate or contains_sky_location but I am open to other suggestions

I guess the question is: what *else* might "contains" make sense for?  The only sensible other thing I could imagine is `contains_pixel`, so that suggests maybe `contains_sky` is sufficient?  TBH I think `contains` is good enough because I think the user wouldn't look to the WCS *anyway* for `contains_pixel` (since that's usually in the image space itself), but I could also see `contains_sky`.

eteq commented:
Oh and re: 
> doesn't having a contains method assume that the data size is part of the public API? 

@ceb8 - This harkens back to a discussion we had on the general WCS API, where it's not required, because you can define a WCS just fine without saying "the edge of the image is pixel #X".  But if this is on the `astropy.wcs.WCS` object, doesn't FITS-WCS actually *require* naxis1/naxis2?

I guess a more general version might be to instead have this method like on `SkyCoord`, like `to_pixel`, and have its signature be `contains(wcs, image=None, ...), where ``image=None`` means "look for naxis1/2", but if the image is given use its `.shape` instead?

bsipocz commented:
> WCS anyway for contains_pixel (since that's usually in the image space itself), but I could also see contains_sky

what about a stacked mosaic camera image? Are the pixels in the gaps in the image or not? (This is purely theoretical question as I guess one wouldn't see that purely looking at the wcs info and not looking for missing values in the data itself.

eteq commented:
> what about a stacked mosaic camera image? Are the pixels in the gaps in the image or not? (This is purely theoretical question as I guess one wouldn't see that purely looking at the wcs info and not looking for missing values in the data itself.

My instinct is to not over-think this too much yet.  I supposed ``image`` could also be a 2-tuple or something that you'd have to calculate by hand in a case like a mosaic, but at that point this "convenience" function is not that convenient.  Or support for this could be added later (in a follow-on PR) by thinking more about what else ``image`` could be.  But I think the functionality @ceb8 added here is still useful for a *lot* of cases.

nden commented:
@eteq It's apparent you don't consider it reasonable to ask if a spectral coordinate (or ra, dec, spec) is contained in a WCS and I wonder why.

As for allowing numerical inputs only - I can see use cases where having to use `SkyCoord` is inconvenience since the WCS internally works with numbers. For example, if I am looking at a combined drizzled image with DS9, I can ask if pixels are within certain individual images or not. The inverse WCS transform has to be used in any case but I don't need to go through `SkyCoord` objects.

These are just general comments, I don't know what the motivation for this work was.
As for implementation, it's a little odd that the `coordinates.SkyCoord.to_pixel` method is used when all it does is call `wcs.utils.skycoord_to_pixel` . Why not use the WCS function directly?

ceb8 commented:
@eteq WRT the private naxis attributes: would it make sense to simply add an error condition if they are not set/0?

@eteq I could certainly hang this off SkyCoord instead of WCS (and call it something like is_in_wcs), this would also sidestep the issue of whether it should take plain numbers and/or arrays.  It felt more natural to me to  hang it on wcs because I was conceptualizing my use-case as "I have a footprint and I need to know if a coordinate is contained in it" but "I have a SkyCoord and I want to know if it's in a WCS footprint" also makes perfect sense.  And maybe putting it there also makes sense since that is where the to_pixel function is.

@nden I did not use `wcs.utils.skycoord_to_pixel` becuase to do so I have to import the utils, which itself imports wcs and that sets up a circular dependency.

@nden The motivation for this work was needing to figure out if a particular sky position was in a wcs footprint I was working with and being surprised that there was not a function that did this, and then thinking other people might also find use from this functionality.

astrofrog commented:
What if this method was called ``footprint_contains`` to make it clear we are talking about the boundaries of the image as a whole, irrespective of data values?

nden commented:
I'm fine with either name. My main issue is whether it should accept numerical only input in addition to SkyCoord objects. I find it useful if it worked with both. However, if others disagree and in the interest of this going forward, if it's considered that it should accept only SkyCoord objects a better name may be `contains_skycoord`?

eteq commented:
> It's apparent you don't consider it reasonable to ask if a spectral coordinate (or ra, dec, spec) is contained in a WCS and I wonder why.

I think it's a reasonable question *also*, but that it's a separate question from what this PR is about.  spectral wcs are a much smaller set of wcs' in the wide world than the spatial ones (at least for the *user* community, which is what this targets). So I'd say the spectral case is useful, but a separate issue.

> accept numerical only input in addition to SkyCoord objects

I am strongly against numerical input. The point of `SkyCoord`/frame classes is to make it so we don't have to support lots of different ways of representing spectral coordinates after all, there should be one (and preferably only one) way to do something.  Also, this way we can ensure the frames are consistent (i.e., the right "kind" of ra/dec).  I guess that's compatible with @nden's last suggestion though: we merge this and have a separate issue to decide on numerical inputs?

As for naming, how about this, a riff on @nden and @astrofrog's:

`SkyCoord.contained(wcs)` <- this is a thin wrapper around the function
`wcs.footprint_contains(sc)` <- this is the function as implemented

I prefer `footprint_contains` over `contains_skycoord` because I think we want this to *also* work with coordinate frame classes, not just skycoords. It may even do so now.  I don't think we need to make it happen for this PR, but if we have "skycoord" in the name that prevents it.  Yet another option to work around that is `wcs.footprint_contains_coord(...)` but that seems overly-verbose.

ceb8 commented:
> SkyCoord.contained(wcs) <- this is a thin wrapper around the function
wcs.footprint_contains(sc) <- this is the function as implemented

I like this solution.  I was also wondering if it makes more sense to hang this functionality off of SkyCoord rather than wcs, but doing that would prevent it being useful for coordinate frame classes.

nden commented:
I was interested in having numerical inputs for performance reasons. Leaving this aside what needs to be done to wrap this PR is to add support for arrays. It should be possible to take as input a `SkyCoord` with a data array and return a boolean array.
@ceb8 Do you have time to finish this?

ceb8 commented:
@nden I'm going to try to finish it up this weekend.

nden commented:
Awesome! Thanks!

ceb8 commented:
Updated the interface as such:
- `SkyCoord.contained_by(wcs, image=None)`
    Moved main functionality here per @nden's comment about the use of skycoord_to_pixel
    image arg makes _naxis optional, if image is present it's dimensions are used instead
- `wcs.footprint_contains(sc)`
    Wrapper around `contained_by`

bsipocz commented:
Given the github outage, CI hasn't yet run on the rebased and finalized PR here, please consider checking for the CI details this before merging.

nden commented:
@ceb8 Sorry if I wasn't clear. We'd like this function to work with SkyCoord objects with data arrays, not just scalars. An example test:
```
ra = np.arange([254.2, 254.1]) 
dec = np.arange([2, 2.1]) 
c = SkyCoord(ra, dec, unit='deg') 
test_wcs.footprint_contains(c)
```
Currently this is failing with
```
ValueError: The truth value of an array with more than one element is ambiguous. 
Use a.any() or a.all()
```
because `contained_by` works with scalars only. It should work with arrays as well.

ceb8 commented:
@nden You were perfectly clear, I just missed that bit when I was going through the changes I needed to make. Should be fixed now.

bsipocz commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1464,7 +1464,7 @@ def contained_by(self, wcs, image=None, **kwargs):
         """
 
-        if image:
+        if not image is None:
```
I think you meant `if image is not None`?

nden commented on astropy/wcs/wcs.py:
```diff
@@ -3084,6 +3084,25 @@ def _as_mpl_axes(self):
+           True means the WCS footprint contains the coordinate, False means it does not.
+        """
+
+        return coord.contained_by(self)
```
Did you mean to pass`**kwargs` here?

ceb8 replied:
Doh....

nden commented:
@ceb8 The Travis flake8 test is failing. It should be a quick fix. Thanks for keeping up with this.

astrofrog commented:
For convenience:

```
astropy/coordinates/sky_coordinate.py:1458:19: W291 trailing whitespace
astropy/coordinates/sky_coordinate.py:1471:1: W293 blank line contains whitespace
astropy/coordinates/sky_coordinate.py:1481:1: W293 blank line contains whitespace
astropy/coordinates/tests/test_sky_coord.py:1532:34: W291 trailing whitespace
astropy/coordinates/tests/test_sky_coord.py:1533:30: W291 trailing whitespace
astropy/coordinates/tests/test_sky_coord.py:1534:43: W291 trailing whitespace
astropy/wcs/wcs.py:3105:19: W291 trailing whitespace
astropy/wcs/tests/test_wcs.py:1162:1: W293 blank line contains whitespace
```

nden commented:
@ceb8 You need to import `warnings` in `sky_coordinate`:
```
>       with warnings.catch_warnings():
E       NameError: name 'warnings' is not defined
astropy/coordinates/sky_coordinate.py:1375: NameError
```

ceb8 commented:
@nden  It's imported right at the top of the file. Is there somewhere else I need to be putting it too?

nden commented:
Interesting. Actually `NoConvergence` is not defined but it's in the `with warnings` block and this shows up as missing `warning` import.  Hm, importing `NoConvergence` from wcs will create a circular import. So perhaps don't specify an exception type, but catch any exception:
```
try:
  ...
except:
  return False
```
If it fails for any reason, then it will be False anyway.

Also there's one more flake8 issue
```
astropy/wcs/wcs.py:3105:19: W291 trailing whitespace
1
```

nden commented:
Argh! flake8 did not like the bare `except`. May be use `except Exception:`

nden commented:
I *think* a rebase will fix the remaining issues. You may want to squash all commits while rebasing.

eteq commented:
It looks like this is passing now so I think it's good to go...

I'm not 100% sure if the last suggestions by @nden were completed, though (since the squashing eliminated that history from this discussion...), so I'll wait a bit for confirmation from @nden (That is, if you're happy feel free to merge it @nden, otherwise I will before the freeze).

nden commented:
Thanks @ceb8 !
